Complexity_classes_def/Cook_Levin.thy

```diff
@@ -0,0 +1,194 @@
+  assumes "poly p_cer"
+  assumes "\<forall>s. \<exists>t. Ex (big_step_t (c, s) t) \<and> t \<le> 
+                      pt (bit_length (s ''input'') + bit_length (s ''certificate''))"
+  assumes "\<forall>x z. \<exists>r. \<forall>s. s ''input'' = x \<and> s ''certificate'' = z \<longrightarrow>
+                         (\<exists>t s'. (c, s) \<Rightarrow>\<^bsup> t \<^esup> s' \<and> s' ''input'' = r)"
+  shows "\<exists>imp_to_sat t_red s_red.
+         poly t_red 
+       \<and> poly s_red
+       \<and> (\<forall>s t s'. (imp_to_sat,s) \<Rightarrow>\<^bsup> t \<^esup> s' \<longrightarrow> s ''certificate'' = s' ''certificate'')
+       \<and> (\<forall>x. \<exists>f.    bit_length f \<le> s_red ( bit_length x )
+                   \<and> (\<forall>s. s ''input'' = x \<longrightarrow> (\<exists>t s'. (imp_to_sat,s) \<Rightarrow>\<^bsup> t \<^esup> s' \<and> 
+                                          s' ''input'' = f \<and> t \<le> t_red(bit_length x)))
+                   \<and> ( f \<in> IMP_SAT \<longleftrightarrow>
+                                        (\<exists>z. bit_length z \<le> p_cer (bit_length x) \<and>
+                                                      (\<forall>s t s'. s ''input'' = x 
+                                                      \<and> s ''certificate'' = z 
+                                                      \<and>(c, s) \<Rightarrow>\<^bsup> t \<^esup> s' \<longrightarrow> s' ''input'' = 0))
+                                        )
+                      )"
+  sorry
+paragraph \<open>Main Lemma on HOL level \<close>
+text \<open>Same lemma but we assume that the reduction is written in HOL \<close>
+lemma main_lemma_hol: 
+  fixes c pt p_cer
+  assumes "poly pt"
+  assumes "poly p_cer"
+  assumes "\<forall>s. \<exists>t. Ex (big_step_t (c, s) t) \<and> t \<le>
+                 pt (bit_length (s ''input'') + bit_length (s ''certificate''))"
+  assumes "\<forall>x z. \<exists>r. \<forall>s. s ''input'' = x \<and> s ''certificate'' = z \<longrightarrow>
+                         (\<exists>t s'. (c, s) \<Rightarrow>\<^bsup> t \<^esup> s' \<and> s' ''input'' = r)"
+   shows "\<exists>imp_to_sat t_red s_red.
+         poly t_red 
+       \<and> poly s_red
+       \<and> (\<forall>x. \<exists>f.    bit_length (encode_sat f) \<le> s_red ( bit_length x ) \<and> imp_to_sat x = f
+                   \<and> ( sat f  \<longleftrightarrow>
+                                        (\<exists>z. bit_length z \<le> p_cer (bit_length x) \<and>
+                                                      (\<forall>s t s'. s ''input'' = x 
+                                                      \<and> s ''certificate'' = z 
+                                                      \<and>(c, s) \<Rightarrow>\<^bsup> t \<^esup> s' \<longrightarrow> s' ''input'' = 0))
+                                        )
+                      )"
+  sorry
+text \<open>Prove for every NP program that it can be reduced to IMP_SAT using 
```

Comment:
problem?

Reply:
it's actually the main-lemma on the HOL level. we should link all the work together to be able to prove it.